python_modules/dagster/dagster/_model/decorator.py

```diff
@@ -0,0 +1,246 @@
+from abc import ABC
+from functools import cached_property, partial
+from typing import (
+    Any,
+    Callable,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Type,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from typing_extensions import dataclass_transform
+
+from dagster._check import EvalContext, build_check_call
+from dagster._utils.cached_method import NT_CACHED_METHOD_CACHE_FIELD
+
+TType = TypeVar("TType", bound=Type)
+TVal = TypeVar("TVal")
+
+
+MODEL_MARKER = object()
+
+
+def _namedtuple_model_transform(
+    cls: TType,
+    *,
+    checked: bool,
+    enable_cached_method: bool,
+    with_new: bool,
+) -> TType:
+    """Transforms the input class in to one that inherits a generated NamedTuple base class
+    and:
+        * bans tuple methods that don't make sense for a model object
+        * creates a run time checked __new__  (optional)
+        * adds @cached_method support (optional).
+    """
+    field_set = {
+        **cls.__annotations__,
+        **({NT_CACHED_METHOD_CACHE_FIELD: Any} if enable_cached_method else {}),
+    }
+    base = NamedTuple(f"_{cls.__name__}", field_set.items())
+
+    if checked:
+        orig_new = base.__new__
+        checks_builder = LazyCheckBuilder(
+            cls.__annotations__,
+            EvalContext.capture_from_frame(2),  # this method, original decorator
+            1 if with_new else 0,
+        )
+
+        def __checked_new__(cls: TType, **kwargs):
+            for key, fn in checks_builder.checks.items():
+                fn(kwargs[key])
+
+            cache_fields = {NT_CACHED_METHOD_CACHE_FIELD: {}} if enable_cached_method else {}
+            return orig_new(cls, **kwargs, **cache_fields)
+
+        base.__new__ = __checked_new__  # type: ignore # unhappy with dropping positional args
+
+    if enable_cached_method:
+        base.__eq__ = __cached_method_safe_eq__
+        base.__hash__ = __cached_method_safe_hash__
+
+    return type(  # type: ignore
+        cls.__name__,
+        (cls, base),
+        {
+            "__iter__": _banned_iter,
+            "__getitem__": _banned_idx,
+            "__hidden_iter__": base.__iter__,
+            "__marker__": MODEL_MARKER,
+            "__annotations__": cls.__annotations__,
+        },
+    )
+
+
+@overload
+def dagster_model(
+    cls: TType,
+) -> TType: ...  # Overload for using decorator with no ().
+
+
+@overload
+def dagster_model(
+    *,
+    checked: bool = True,
+    enable_cached_method: bool = False,
+) -> Callable[[TType], TType]: ...  # Overload for using decorator used with args.
+
+
+@dataclass_transform(
+    kw_only_default=True,
+    frozen_default=True,
+)
+def dagster_model(
+    cls: Optional[TType] = None,
```

Comment:
I think this `= None` the crux of my confusion when it comes to understanding why we can't have a `with_new` boolean. Under what circumstances is this None?

Reply:
its the decorator factory pattern ish - so 
* `@dagster_model` - gets the class as positional arg 
* `@dagster_model(checked=False)` - no positional, uses None cls 

The whole "with_new" thing exists because we use `@dataclass_transform` to tell static analysis that this decorator turns the class in to a dataclass-esque thing with a constructor and access based on the field set.

`@dataclass_transform` doesn't follow if you override `__new__`, so static analysis expects the field set based construction instead of what you overrode to. Since you are declaring `__new__` though, you don't need `@dataclass_transform` so we just need to not apply that if you are specifying `__new__`. 

Ideally we could do this by applying `@dataclass_transform` to `@overloads` where `with_new` arg is not `Literal[true]`, but unfortunately when `@dataclass_transform` is used it applies to all declarations of the function. From https://peps.python.org/pep-0681/

> When applied to an overload, the dataclass_transform decorator still impacts all usage of the function.

So the only way to from what i can tell is have one variant of the function with `@dataclass_transform` and one without.

Reply:
Cool this all makes sense thanks for writing it out.

This might be worth encoding as a comment.